% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/annotation_gff.R
\name{pattern_count_genome}
\alias{pattern_count_genome}
\title{Find how many times a given pattern occurs in every gene of a genome.}
\usage{
pattern_count_genome(
  fasta,
  gff = NULL,
  pattern = "TA",
  type = "gene",
  id_col = "ID",
  key = NULL
)
}
\arguments{
\item{fasta}{Genome sequence.}

\item{gff}{Gff of annotation information from which to acquire CDS (if not
provided it will just query the entire genome).}

\item{pattern}{What to search for? This was used for tnseq and TA is the
mariner insertion point.}

\item{type}{Column to use in the gff file.}

\item{id_col}{Column containing the gene IDs.}

\item{key}{What type of entry of the gff file to key from?}
}
\value{
Data frame of gene names and number of times the pattern appears/gene.
}
\description{
There are times when knowing how many times a given string appears in a
genome/CDS is helpful. This function provides that information and is
primarily used by cp_seq_m().
}
\details{
This is once again a place where mcols() usage might improve the overall
quality of life.
}
\examples{
 pa_data <- get_paeruginosa_data()
 pa_fasta <- pa_data[["fasta"]]
 pa_gff <- pa_data[["gff"]]
 ta_count <- pattern_count_genome(pa_fasta, pa_gff)
 head(ta_count)
}
\seealso{
[Biostrings] [Rsamtools::FaFile()] [Biostrings::PDict()]
}
