% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/de_plots.R
\name{plot_ma_de}
\alias{plot_ma_de}
\title{Make a pretty MA plot from one of limma, deseq, edger, or basic.}
\usage{
plot_ma_de(
  table,
  expr_col = "logCPM",
  fc_col = "logFC",
  p_col = "qvalue",
  pval = 0.05,
  alpha = 0.4,
  logfc = 1,
  label_numbers = TRUE,
  size = 2,
  shapes = TRUE,
  invert = FALSE,
  label = NULL,
  label_column = "hgncsymbol",
  ...
)
}
\arguments{
\item{table}{Df of linear-modelling, normalized counts by sample-type,}

\item{expr_col}{Column showing the average expression across genes.}

\item{fc_col}{Column showing the logFC for each gene.}

\item{p_col}{Column containing the relevant p values.}

\item{pval}{Name of the pvalue column to use for cutoffs.}

\item{alpha}{How transparent to make the dots.}

\item{logfc}{Fold change cutoff.}

\item{label_numbers}{Show how many genes were 'significant', 'up', and 'down'?}

\item{size}{How big are the dots?}

\item{shapes}{Provide different shapes for up/down/etc?}

\item{invert}{Invert the ma plot?}

\item{label}{Label the top/bottom n logFC values?}

\item{label_column}{gene annotation column from which to extract labels.}

\item{...}{More options for you}
}
\value{
ggplot2 MA scatter plot.  This is defined as the rowmeans of the
 normalized counts by type across all sample types on the x axis, and the
 log fold change between conditions on the y-axis. Dots are colored
 depending on if they are 'significant.'  This will make a fun clicky
 googleVis graph if requested.
}
\description{
Because I can never remember, the following from wikipedia: "An MA plot is an
application of a Bland-Altman plot for visual representation of two channel
DNA microarray gene expression data which has been transformed onto the M
(log ratios) and A (mean average) scale."
}
\examples{
 \dontrun{
  plot_ma(voomed_data, table)
  ## Currently this assumes that a variant of toptable was used which
  ## gives adjusted p-values.  This is not always the case and I should
  ## check for that, but I have not yet.
 }
}
\seealso{
[limma_pairwise()] [deseq_pairwise()] [edger_pairwise()] [basic_pairwise()]
}
